dev/tools/lib/repository.dart

```diff
@@ -0,0 +1,357 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert' show jsonDecode;
+import 'dart:io' as io;
+
+import 'package:file/file.dart';
+import 'package:meta/meta.dart';
+import 'package:process/process.dart';
+import 'package:platform/platform.dart';
+
+import './git.dart';
+import './globals.dart' as globals;
+import './stdio.dart';
+import './version.dart';
+
+/// A source code repository.
+class Repository {
+  Repository({
+    @required this.name,
+    @required this.upstream,
+    @required this.processManager,
+    @required this.stdio,
+    @required this.platform,
+    @required this.fileSystem,
+    @required this.parentDirectory,
+    this.localUpstream = false,
+    this.useExistingCheckout = false,
+  })  : git = Git(processManager),
+        assert(localUpstream != null),
+        assert(useExistingCheckout != null);
+
+  final String name;
+  final String upstream;
+  final Git git;
+  final ProcessManager processManager;
+  final Stdio stdio;
+  final Platform platform;
+  final FileSystem fileSystem;
+  final Directory parentDirectory;
+  final bool useExistingCheckout;
+
+  /// If the repository will be used as an upstream for a test repo.
+  final bool localUpstream;
+
+  Directory _checkoutDirectory;
+
+  /// Lazily-loaded directory for the repository checkout.
+  ///
+  /// Cloning a repository is time-consuming, thus the actual repository is not
+  /// ensured to exist until this getter is called.
+  Directory get checkoutDirectory {
+    if (_checkoutDirectory == null) {
```

Comment:
you could un-nest this with an early return:

```
if (_checkoutDirectory != null) {
  ...
}
```

Reply:
i think rubocop (ruby's linter) favors as few returns as possible, but I agree with your comment that as few indents as possible is favorable.